### PR TITLE
Refactor Issues Subscription

### PR DESCRIPTION
**Pull request by 6543**
* [x] FIX: getIssueWatchers() get only aktive suscriber

* [x] update tests
* [x] add tests

* [ ] ~~ignore issue if unsubscribed but repowatch~~ (Moved to #8852)

### CODE REVIEW

File: models/issue_watch.go [hunk 1]
```diff
@@ -68,10 +69,16 @@ func GetIssueWatchers(issueID int64) (IssueWatchList, error) {
 func getIssueWatchers(e Engine, issueID int64) (watches IssueWatchList, err error) {
 	err = e.
 		Where("`issue_watch`.issue_id = ?", issueID).
+		And("`issue_watch`.is_watching = ?", true).
 		And("`user`.is_active = ?", true).
 		And("`user`.prohibit_login = ?", false).
 		Join("INNER", "`user`", "`user`.id = `issue_watch`.user_id").
 		Find(&watches)
+	// handle default watchers (repowatchers)
```
Comment (6543) [inline models/issue_watch.go]: dont add this jet - implement different watching level first

**Comment by 6543** [comment]
move ToDos into new issue: #8852 

to make refactor independend from feature additions

File: models/fixtures/issue_watch.yml [hunk 2]
```diff
@@ -13,3 +13,19 @@
...
+-
+  id: 3
+  user_id: 2
+  issue_id: 7
+  is_watching: true
+  created_unix: 946684800
+  updated_unix: 946684800
+
+-
+  id: 3
```
Comment (6543) [inline models/fixtures/issue_watch.yml]: id is unice!!

**Comment by 6543** [comment]
@lunny I think it is ready for review ...

File: models/issue_watch.go [hunk 3]
```diff
@@ -21,7 +21,21 @@ type IssueWatchList []*IssueWatch
 
 // CreateOrUpdateIssueWatch set watching for a user and issue
 func CreateOrUpdateIssueWatch(userID, issueID int64, isWatching bool) error {
-	iw, exists, err := getIssueWatch(x, userID, issueID)
+	sess := x.NewSession()
+	defer sess.Close()
+
+	if err := sess.Begin(); err != nil {
```
Comment (guillep2k) [inline models/issue_watch.go]: Why do we need a separate function? As long as `createOrUpdateIssueWatch()` performs only one statement (either insert or update), IMHO there no need for a transaction here. Unless you need to call `createOrUpdateIssueWatch()` as part of a bigger transaction from other parts of the code.

File: models/issue_watch_test.go [hunk 4]
```diff
@@ -30,7 +30,7 @@ func TestGetIssueWatch(t *testing.T) {
 	assert.NoError(t, err)
 
 	_, exists, err = GetIssueWatch(2, 2)
-	assert.Equal(t, true, exists)
+	assert.Equal(t, false, exists)
```
Comment (guillep2k) [inline models/issue_watch_test.go]: `assert.False(t, ....`

File: models/issue_watch_test.go [hunk 5]
```diff
@@ -48,9 +48,15 @@ func TestGetIssueWatchers(t *testing.T) {
...
+	// Watcher is not watching
+	assert.Equal(t, 0, len(iws))
 
 	iws, err = GetIssueWatchers(5)
 	assert.NoError(t, err)
 	assert.Equal(t, 0, len(iws))
+
+	iws, err = GetIssueWatchers(7)
+	assert.NoError(t, err)
+	// Watcher is not watching
```
Comment (guillep2k) [inline models/issue_watch_test.go]: Comment is not accurate

**Review by guillep2k — COMMENTED** [review]
As for the API I'm not familiar with it, so I can't emit an opinion about that part. It looks like there's a few changes.

File: models/issue_watch.go (hunk 3 above)
Comment (6543) [reply models/issue_watch.go]: ok ... I'll revert c1de540

**Comment by 6543** [comment]
@guillep2k there is only one bugfix (dont return watcher witch is explicite NOT watching) the rest is only code refactoring + add tests

File: routers/api/v1/api.go [hunk 6]
```diff
@@ -691,9 +691,9 @@ func RegisterRoutes(m *macaron.Macaron) {
 							m.Post("/stop", reqToken(), repo.StopIssueStopwatch)
 						})
 						m.Group("/subscriptions", func() {
-							m.Get("", bind(api.User{}), repo.GetIssueSubscribers)
-							m.Put("/:user", repo.AddIssueSubscription)
-							m.Delete("/:user", repo.DelIssueSubscription)
+							m.Get("", repo.GetIssueSubscribers)
```
Comment (guillep2k) [inline routers/api/v1/api.go:694]: Why doesn't this need `reqToken()`?

Comment (6543) [reply routers/api/v1/api.go]: @guillep2k 
It set/unset subscribtion - and currently the only check is ctx.user == user || ctx.user.is_admin
with token req it should be more secure ?

**Comment by 6543** [comment]
@guillep2k add reqToken to GET too? or how open/restrictive this function should be?

**Comment by guillep2k** [comment]
It's no longer leaking the mail addresses through this endpoint. However, there's no longer a way for admins to get the users' addresses through most endpoints including this (`/admin/adminGetAllUsers` seems to work). I don't know, looks like a breaking change?

**Comment by 6543** [comment]
We could Introduce an User.AdminAPIFormat() func ? @guillep2k  wich is exec if user is admin?

**Comment by 6543** [comment]
@guillep2k the PR dont change API behaviour only fix a bug and refactor code ...

**Comment by 6543** [comment]
@lunny I'm ready ...

**Review by zeripath — APPROVED** [review]

**Comment by lunny** [comment]
Make L-G-T-M work

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by guillep2k** [comment]
> force-pushed the 6543-forks:refactor_issues-subscription branch 2 times, most recently from 37d0624 to 0ac149a 2 days ago 

@6543 first force-push fixed the user email leaking, second force-push removed the fix? I can't find that fix anymore.

**Comment by 6543** [comment]
@guillep2k I removed this "fix" - because there was no bug user.apiformat already handle this for us ...

File: routers/api/v1/api.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (6543) [reply routers/api/v1/api.go:694]: @lunny because we dont need/use it ... (useless for GET)

Comment (6543) [reply routers/api/v1/api.go:694]: m.Get("/subscriptions", user.GetWatchedRepos) also dont require Token ...

Comment (6543) [reply routers/api/v1/api.go:694]: @guillep2k should I add `reqToken()` ?!?

or can I "resolve" this for now?!?

**Comment by 6543** [comment]
@lunny can you look at https://github.com/go-gitea/gitea/pull/8738#discussion_r347146877 plz

**Review by lunny — APPROVED** [review]

**Pull request merged**